### PR TITLE
Implement initial version of data clustering CLI tool

### PR DESCRIPTION
**Pull request by kirillsakh**
Task Overview

Develop a command-line tool for data clustering.

Objectives

- Code Quality: Code should be clear and well-organized, exemplifying the essence of durable software.
- Best Practices: Follow industry-standard best practices, including SOLID principles, to ensure the software remains scalable and of high quality.
- Dependency injection: Leverage the Dependency Injector library for efficient management of software components.
- Style Consistency: Set up pre-commit hooks to automate checks for code quality, including syntax verification, linting, and formatting.

Goal

Demonstrate technical skills, dedication to quality, and comprehensive approach to software development from the initial concept to the final implementation.

Specifications

- Input Data: The tool should accept data in either JSON format or as NumPy arrays.
- Output Data: Ensure the tool supports output in both JSON format and NumPy arrays.
- Clustering:
  - Framework: Utilize scikit-learn for all clustering-related functionalities.
  - Algorithm Selection: Select and implement three distinct clustering algorithms from scikit-learn to illustrate varied strategies.
  - Hyperparameters: Provide options for customizable hyperparameter settings.
- Configuration: Manage all settings, including choices of algorithms, hyperparameters, and data formats, through a config.yaml file. Use providers.Selector for flexible algorithm selection, enhancing the tool’s adaptability.

Code and Repository Requirements

- Readability and Structure: Ensure code is readable and elegantly structured.
- Documentation: Thoroughly document methods and functions using google style docstrings.
- Versioning: maintain semantic version of the tool
- Testing: Perform unit tests, and, where applicable, end-to-end or integration tests utilizing pytest and mocker.
- Pre-commit Setup: Configure pre-commit hooks for automated checks to maintain code quality.

Git Practices

Maintain an organized commit history by making small, self-contained commits that either introduce a single change or add a new feature.

### CODE REVIEW

**Comment by kirillsakh** [comment]
My idea is to make this tool as installable package via tar.gz file. WDYT?

**Comment by kirillsakh** [comment]
One of the suggestions you made was:

Utilize `providers.Selector` to enable flexible algorithm selection, thereby improving the adaptability of the tool.

Could you elaborate on how `poviders.Selector` could be integrated into the current implementation of the clustering tool?

File: config.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+default_format: json
+
+formats:
+  json: src.file_handlers.json_handler.JsonHandler
+  npy: src.file_handlers.numpy_handler.NumpyHandler
+
+default_algorithm: kmeans
+
+algorithms:
```
Comment (tynaneu) [inline config.yaml:9]: this should not be here - this should be in di (config.py)

File: config.yaml [hunk 2]
```diff
@@ -0,0 +1,22 @@
+default_format: json
+
+formats:
```
Comment (tynaneu) [inline config.yaml:3]: this should not be here - this should be in di (config.py)

File: config.yaml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  json: src.file_handlers.json_handler.JsonHandler
+  npy: src.file_handlers.numpy_handler.NumpyHandler
+
+default_algorithm: kmeans
```
Comment (tynaneu) [inline config.yaml:7]: can you move it to section with hyperparameters as "type" and rename the section "hyperparameters" to "clustering"

File: config.yaml [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+algorithms:
+  kmeans: sklearn.cluster.KMeans
+  spectral: sklearn.cluster.SpectralClustering
+  dbscan: sklearn.cluster.DBSCAN
+
+hyperparams:
```
Comment (tynaneu) [inline config.yaml:14]: please rename the section to "clustering" and use "type" for selecting the algorithm

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -0,0 +1,18 @@
+repos:
+  - repo: https://github.com/pycqa/isort
+    # isort version.
+    rev: 5.11.2
+    hooks:
+      # Run the isort.
+      - id: isort
```
Comment (tynaneu) [inline .pre-commit-config.yaml:7]: isort is part of ruff - so it is duplicate

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -0,0 +1,18 @@
+repos:
+  - repo: https://github.com/pycqa/isort
+    # isort version.
+    rev: 5.11.2
+    hooks:
```
Comment (tynaneu) [inline .pre-commit-config.yaml:5]: please add mypy

File: src/cluster_crafter.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    If the `method` parameter is not provided, the default clustering algorithm specified in the `config.yaml` file will be used.
+
+    If the `output_file` parameter is not provided, the clustering labels will be saved into a file named `labels` with the default file format specified in the `config.yaml` file.
+
+
+    Example usage:
+        python cli.py --input-file=data.json --method=kmeans --output-file=clustered_data.json
+    """
+
+    # Create a container for configuration settings
```
Comment (tynaneu) [inline src/cluster_crafter.py:34]: please use comments only if it is not straightforward from code

File: src/cluster_crafter.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    config_container = ConfigContainer.create(config_file=get_config_file())
+
+    # Load data from input file
+    data = _load_data(config_container, input_file)
+
+    # Perform clustering
+    labels = _process_data(config_container, data, method)
```
Comment (tynaneu) [inline src/cluster_crafter.py:41]: clustering should be encapsulated in class to use it in dependency injection in selectors (see https://python-dependency-injector.ets-labs.org/providers/selector.html)

**Pull request closed without merging**